Video-Transformers-Kotlin/app/src/main/java/com/tokbox/sample/basicvideochat/MainActivity.kt

```diff
@@ -0,0 +1,325 @@
+                    initializeSession(it.apiKey, it.sessionId, it.token)
+                }
+            }
+
+            override fun onFailure(call: Call<GetSessionResponse?>, t: Throwable) {
+                throw RuntimeException(t.message)
+            }
+        })
+    }
+
+    private fun initializeSession(apiKey: String, sessionId: String, token: String) {
+        Log.i(TAG, "apiKey: $apiKey")
+        Log.i(TAG, "sessionId: $sessionId")
+        Log.i(TAG, "token: $token")
+
+        /*
+        The context used depends on the specific use case, but usually, it is desired for the session to
+        live outside of the Activity e.g: live between activities. For a production applications,
+        it's convenient to use Application context instead of Activity context.
+         */
+        session = Session.Builder(this, apiKey, sessionId).build().also {
+            it.setSessionListener(sessionListener)
+            it.connect(token)
+        }
+    }
+
+    private fun initRetrofit() {
+        val logging = HttpLoggingInterceptor()
+        logging.setLevel(HttpLoggingInterceptor.Level.BODY)
+        val client: OkHttpClient = OkHttpClient.Builder()
+            .addInterceptor(logging)
+            .build()
+
+        retrofit = Retrofit.Builder()
+            .baseUrl(ServerConfig.CHAT_SERVER_URL)
+            .addConverterFactory(MoshiConverterFactory.create())
+            .client(client)
+            .build().also {
+                apiService = it.create(APIService::class.java)
+            }
+    }
+
+    private fun finishWithMessage(message: String) {
+        Log.e(TAG, message)
+        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
+        finish()
+    }
+
+    companion object {
+        private val TAG = MainActivity::class.java.simpleName
+        private const val PERMISSIONS_REQUEST_CODE = 124
+    }
+
+    class logoWatermark(private val resources: Resources) : CustomVideoTransformer {
+
+        // Get the image in bitmap format
+        private var image: Bitmap = BitmapFactory.decodeResource(resources, R.drawable.vonage_logo)
+
+        fun resizeImage(image: Bitmap?, width: Int, height: Int): Bitmap {
+            return Bitmap.createScaledBitmap(image!!, width, height, true)
+        }
+
+        override fun onTransform(frame: BaseVideoRenderer.Frame) {
+
+            // Obtain the Y  plane of the video frame
+            val yPlane: ByteBuffer = frame.yplane
+
+            // Get the dimensions of the video frame
+            val videoWidth = frame.width
+            val videoHeight = frame.height
+
+            // Calculate the desired size of the image
+            val desiredWidth = videoWidth / 8 // Adjust this value as needed
+            val desiredHeight: Int = (image.getHeight() * (desiredWidth.toFloat() / image.getWidth())).toInt()
+
+            // Resize the image to the desired size
+            image = resizeImage(image, desiredWidth, desiredHeight)
```

Comment:
What's the point in updating image (the member object) in every iteration? It would seem some kind of cache... but actually it is not. There is another issue: this works in only one direction. Every call to resize is based on a previous call. What if a 2x2 frame appears? How can it be resized later? (I know lowest resolution is 320x240 though)

Reply:
Some several reasons, during runtime the resolution will change. On startup for example the resolution usually drops and goes back to normal. I had to create a function to resize the image because if not it would seem the image size was changing but it is actually the video size that is is changing. 
I am uptating on every iteraction, but I can check if the resolution was changed, and only in that case, the image is updated. Relating to the resizing only working on one way, I tested it out and it works well in case the value increases or decreases

I will change so that we only resize if size changes

Reply:
Yes, it work both ways, but I meant once it is resized down, some details are lost forever. In any case, probably not relevant for such an image :)

Now I am thinking that you don't need to check. Actually I guess the call to createScaledBitmap is a no-op if the resolution doesn't change.

Reply:
You are right. The image can loose quality when converting from lower to higher resolution. The image is no immutable, and we use a second variable to resize it if needed